crates/deltalake-core/src/operations/constraints.rs

```diff
@@ -0,0 +1,263 @@
+
+            if metadata.configuration.contains_key(&configuration_key) {
+                return Err(DeltaTableError::Generic(format!(
+                    "Constraint with name: {} already exists, expr: {}",
+                    name, expr
+                )));
+            }
+
+            let state = this.state.unwrap_or_else(|| {
+                let session = SessionContext::new();
+                register_store(this.log_store.clone(), session.runtime_env());
+                session.state()
+            });
+
+            // Checker built here with the one time constraint to check.
+            let checker = DeltaDataChecker::with_constraints(vec![Constraint::new("*", &expr)]);
+
+            let files_to_check =
+                find_files(&this.snapshot, this.log_store.clone(), &state, None).await?;
+            let scan = DeltaScanBuilder::new(&this.snapshot, this.log_store.clone(), &state)
+                .with_files(&files_to_check.candidates)
+                .build()
+                .await?;
+
+            let task_ctx = Arc::new(TaskContext::from(&state));
+            let record_stream: SendableRecordBatchStream = scan.execute(0, task_ctx)?;
+            let records = collect_sendable_stream(record_stream).await?;
+
+            for batch in records {
+                checker.check_batch(&batch).await?;
+            }
+
+            // We have validated the table passes it's constraints, now to add the constraint to
+            // the table.
+
+            metadata
+                .configuration
+                .insert(format!("delta.constraints.{}", name), Some(expr.clone()));
+
+            let old_protocol = this.snapshot.protocol();
+            let protocol = Protocol {
+                min_reader_version: if old_protocol.min_reader_version > 1 {
+                    old_protocol.min_reader_version
+                } else {
+                    1
+                },
+                min_writer_version: if old_protocol.min_reader_version > 3 {
+                    old_protocol.min_reader_version
+                } else {
+                    3
+                },
+                reader_features: old_protocol.reader_features.clone(),
+                writer_features: old_protocol.writer_features.clone(),
+            };
+
+            let operational_parameters = HashMap::from_iter([
+                ("name".to_string(), json!(&name)),
+                ("expr".to_string(), json!(&expr)),
+            ]);
+            let commit_info = CommitInfo {
+                timestamp: Some(Utc::now().timestamp_millis()),
+                operation: Some("ADD CONSTRAINT".to_string()),
+                operation_parameters: Some(operational_parameters),
+                read_version: Some(this.snapshot.version()),
+                isolation_level: Some(IsolationLevel::Serializable),
+                is_blind_append: Some(false),
+                ..Default::default()
+            };
```

Comment:
We usually create the `CommitInfo` via the assoaciated `commit_info` method on the `DeltaOperation`.

Reply:
This doesn't seem better than what I'm doing here, I'd still have to set most of this myself. I took this info from the spark implementation when I created it. Is that okay?

Reply:
Yeah its fine .. the creation of the commit info itself is not so critical.

The only really valuable thing over there is that we have a single place where we can define the string representation for an operation, in this case `ADD CONSTRAINT`. In erarlier version of delta-rs we hade some inconsistencies in that regard.

Come to think of it, most information - like `read_version`, `isolation_level` etc. we should probably set in a central place within the `commit` function anyhow. Especially since there are some cases where we can (and do) relax the isolation level for the actual commit. Should we maybe add an issue for that?